### PR TITLE
feat: Query Available for Borrow

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Add structures to retrieve the Amount available for borrowing of a specific denom

ref: #343 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by toteki** [comment]
In this case you can use `sdk.Int` instead of `sdk.Dec` for the response field, similar to how `QueryReserveAmount` works

File: x/leverage/keeper/grpc_query.go [hunk 1]
```diff
@@ -94,6 +94,29 @@ func (q Querier) Borrowed(
...
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if !q.Keeper.IsAcceptedToken(ctx, req.Denom) {
+		return nil, status.Error(codes.InvalidArgument, "not accepted Token denom")
+	}
+
+	// Available for borrow = Module Balance - Reserve Amount
+	moduleBalance := q.Keeper.ModuleBalance(ctx, req.Denom)
+	reserveAmount := q.Keeper.GetReserveAmount(ctx, req.Denom)
```
Comment (alexanderbez) [inline x/leverage/keeper/grpc_query.go:115]: nit: can we actually put this into a method on the `keeper` type? You never know where else we might need this.

**Review by alexanderbez — CHANGES_REQUESTED** [review]

**Comment by RafilxTenfen** [comment]
> In this case you can use `sdk.Int` instead of `sdk.Dec` for the response field, similar to how `QueryReserveAmount` works

done at https://github.com/umee-network/umee/pull/379/commits/ff5848bd1ba812b583770dbb71ccc0bf5d2457d5

File: x/leverage/keeper/grpc_query.go (hunk 1 above)
Comment (RafilxTenfen) [reply x/leverage/keeper/grpc_query.go]: done at https://github.com/umee-network/umee/pull/379/commits/5692bbea45ec0508050569f9a4fc947f731872b2

File: x/leverage/keeper/token.go [hunk 2]
```diff
@@ -197,3 +197,12 @@ func (k Keeper) GetLiquidationIncentive(ctx sdk.Context, denom string) (sdk.Dec,
...
 }
+
+// GetAvailableToBorrow gets the amount available to borrow of a given token.
+func (k Keeper) GetAvailableToBorrow(ctx sdk.Context, denom string) sdk.Int {
+	// Available for borrow = Module Balance - Reserve Amount
+	moduleBalance := k.ModuleBalance(ctx, denom)
+	reserveAmount := k.GetReserveAmount(ctx, denom)
+
+	return moduleBalance.Sub(reserveAmount)
+}
```
Comment (toteki) [inline x/leverage/keeper/token.go:208]: This can move to `x/leverage/keeper/borrows.go` where similar keeper methods are

Comment (RafilxTenfen) [reply x/leverage/keeper/token.go]: done at https://github.com/umee-network/umee/pull/379/commits/222e5b97db09336eba8affdf267766c4f6fdf73f

File: x/leverage/keeper/borrows.go [hunk 3]
```diff
@@ -210,3 +210,12 @@ func (k Keeper) SetBorrowAPY(ctx sdk.Context, denom string, borrowAPY sdk.Dec) e
 	store.Set(key, bz)
 	return nil
 }
+
+// GetAvailableToBorrow gets the amount available to borrow of a given token.
+func (k Keeper) GetAvailableToBorrow(ctx sdk.Context, denom string) sdk.Int {
+	// Available for borrow = Module Balance - Reserve Amount
+	moduleBalance := k.ModuleBalance(ctx, denom)
+	reserveAmount := k.GetReserveAmount(ctx, denom)
+
```
Comment (alexanderbez) [inline x/leverage/keeper/borrows.go:219]: @toteki do we need any zero amount defense checks here? What if `moduleBalance < reserveAmount`?

File: x/leverage/keeper/borrows.go [hunk 4]
```diff
@@ -210,3 +210,12 @@ func (k Keeper) SetBorrowAPY(ctx sdk.Context, denom string, borrowAPY sdk.Dec) e
...
+	return moduleBalance.Sub(reserveAmount)
```
Comment (toteki) [inline x/leverage/keeper/borrows.go:220]:
```suggestion
	return sdk.MaxInt(moduleBalance.Sub(reserveAmount),sdk.ZeroInt())
```

File: x/leverage/keeper/borrows.go (hunk 3 above)
Comment (toteki) [reply x/leverage/keeper/borrows.go:219]: Added in batch

**Comment by toteki** [comment]
Piggybacked a one-line change to reserves.go (obscure bugfix) with my last commit

**Comment by alexanderbez** [comment]
Can we get tests and linting addressed and then we can merge this 🙏

**Pull request merged**